### ISSUE TITLE
Best way to split multi-task datasets with partially missing endpoint values using Hestia?

### ISSUE THREAD

**Issue by mmagithub** [IBM/Hestia-GOOD#95]
<p data-start="262" data-end="265">Hi,</p><p data-start="267" data-end="351">Thanks again for this excellent library—it's been incredibly useful in our projects.</p><p data-start="353" data-end="748">I'm looking for advice on how Hestia can handle data splitting for multi-task learning scenarios. Specifically, we have datasets where multiple properties (numerical regression targets or binary classification labels) are associated with each molecule (represented by SMILES). However, not all molecules have values available for all endpoints (some endpoints are missing for certain molecules).</p><p data-start="750" data-end="783"><strong data-start="750" data-end="783">What we're trying to achieve:</strong></p><p data-start="785" data-end="1112">We want to split the dataset into training, validation, and test sets in a way that ensures each subset maintains a representative distribution across all properties, even though some property values may be partially missing. The goal is to maximize the chance of obtaining uniform and meaningful distributions for each subset.</p><p data-start="1114" data-end="1137"><strong data-start="1114" data-end="1137">Simple Toy Example:</strong></p><p data-start="1139" data-end="1176">Consider the following small dataset:</p>

| SMILES     | Property_A | Property_B | Property_C |
|------------|------------|------------|------------|
| CCO        | 1.2        | 0          | NaN        |
| CCN        | NaN        | 1          | 3.5        |
| CCCO       | 2.3        | NaN        | NaN        |
| CC(=O)O    | 1.5        | 0          | 2.2        |
| CCN(CC)CC  | NaN        | 1          | NaN        |
| C1CCCCC1   | 3.1        | 0          | 4.2        |
| COC        | NaN        | NaN        | 1.8        |
| CC(C)O     | 2.2        | 1          | NaN        |

<p data-start="1739" data-end="1755">In this example:</p><ul data-start="1757" data-end="1939"><li data-start="1757" data-end="1817">Each row corresponds to a molecule (identified by SMILES).</li><li data-start="1818" data-end="1896"><code data-start="1820" data-end="1832">Property_A</code>, <code data-start="1834" data-end="1846">Property_B</code>, and <code data-start="1852" data-end="1864">Property_C</code> are tasks we want to predict.</li><li data-start="1897" data-end="1939"><code data-start="1899" data-end="1904">NaN</code> indicates missing property values.</li></ul><p data-start="1941" data-end="1959"><strong data-start="1941" data-end="1959">Our challenge:</strong></p><ul data-start="1961" data-end="2362"><li data-start="1961" data-end="2111">A naive approach would be to randomly split all molecules into subsets. However, this risks subsets having uneven representations for each property.</li><li data-start="2112" data-end="2219">Alternatively, splitting each task independently might yield inconsistent subsets that are hard to merge.</li><li data-start="2220" data-end="2362">Ideally, we'd like a unified splitting strategy that considers all tasks simultaneously, resulting in balanced distributions across subsets.</li></ul><p data-start="2364" data-end="2554">An actual use case with a more complex scenario is the Biogen dataset, which follows this type of structure, but we're also interested in applying the strategy to other large-scale datasets.</p><p data-start="2556" data-end="2569"><strong data-start="2556" data-end="2569">Question:</strong></p><p data-start="2571" data-end="2759">Is there a built-in solution or recommended workaround in Hestia to achieve a unified and representative splitting strategy in the presence of partially missing multi-task endpoint values?</p><p data-start="2761" data-end="2783" data-is-last-node="" data-is-only-node="">Many thanks,<br data-start="2773" data-end="2776">Marawan</p>

**Comment by RaulFD-creator** [comment]
Hi Marawan,

Thanks for the kind words! I'm glad that you find it useful. The first step would be to create value bins, so that the algorithm can stratify based on the bin (the stratification, currently does not support regression targets). For example, using the`KBinsDiscretizer` from `sklearn`. This categoricalization would not address the problem of missing endpoints as NaN values, so they should be assigned to a category of their own, either before or after.  Once you do that, it may work with the current `ccpart` algorithm, you just need to specify the `label_field` column in the input dataframe. However, I've not tested it. If you encounter problems, please let me know and  I could try running some tests to see what parts of the code might need generalising.

Kind regards,
Raul
